pkg/operator/startupmonitorreadiness/readiness_checks.go

```diff
@@ -0,0 +1,321 @@
+	revisionString, found := kasPod.Labels["revision"]
+	if !found {
+		return false, "InvalidPod", fmt.Sprintf("pod %s doesn't have revision label", kasPod.Name)
+	}
+	if len(revisionString) == 0 {
+		return false, "InvalidPod", fmt.Sprintf("empty revision label on %s pod", kasPod.Name)
+	}
+	revision, err := strconv.Atoi(revisionString)
+	if err != nil || revision < 0 {
+		return false, "InvalidPod", fmt.Sprintf("invalid revision label on pod %s: %q", kasPod.Name, revisionString)
+	}
+
+	if revision != monitorRevision {
+		return false, "InvalidPod", fmt.Sprintf("the running Kube API is at unexpected revision %d, expected %d", revision, monitorRevision)
+	}
+
+	return true, "", ""
+}
+
+// goodReadyzEndpoint performs HTTP checks against readyz?verbose=true endpoint
+//  returns true, "", "", when we got HTTP 200 "successThreshold" times
+//  returns false, "NotReady", EntireResponseBody (if any) on HTTP != 200
+//  returns false, "NotReadyError", EntireResponseBody (if any) in case of any error or timeout
+func goodReadyzEndpoint(client *http.Client, rawURL string, successThreshold int, interval time.Duration) func(ctx context.Context) (bool, string, string) {
+	return func(ctx context.Context) (bool, string, string) {
+		return doHTTPCheckAndTransform(ctx, client, fmt.Sprintf("%s/readyz?verbose=true", rawURL), "NotReady", doHTTPCheckMultipleTimes(successThreshold, interval))
+	}
+}
+
+// goodHealthzEndpoint performs an HTTP check against healthz?verbose=true endpoint
+//  returns true, "", "", on HTTP 200
+//  returns false, "Unhealthy", EntireResponseBody (if any) on HTTP != 200
+//  returns false, "UnhealthyError", EntireResponseBody (if any) in case of any error or timeout
+func goodHealthzEndpoint(client *http.Client, rawURL string) func(context.Context) (bool, string, string) {
+	return func(ctx context.Context) (bool, string, string) {
+		return doHTTPCheckAndTransform(ctx, client, fmt.Sprintf("%s/healthz?verbose=true", rawURL), "Unhealthy", doHTTPCheck)
+	}
+}
+
+// goodETCDEndpoint performs an HTTP check against healthz/etcd endpoint
+//  returns true, "", "", on HTTP 200
+//  returns false, "EtcdUnhealthy", EntireResponseBody (if any) on HTTP != 200
+//  returns false, "EtcdUnhealthyError", EntireResponseBody (if any) in case of any error or timeout
+func goodETCDEndpoint(client *http.Client, rawURL string) func(context.Context) (bool, string, string) {
+	return func(ctx context.Context) (bool, string, string) {
+		return doHTTPCheckAndTransform(ctx, client, fmt.Sprintf("%s/healthz/etcd", rawURL), "EtcdUnhealthy", doHTTPCheck)
+	}
+}
+
+func doHTTPCheckAndTransform(ctx context.Context, client *http.Client, rawURL string, checkName string, httpCheckFn func(ctx context.Context, client *http.Client, rawURL string) (int, string, error)) (bool, string, string) {
+	statusCode, response, err := httpCheckFn(ctx, client, rawURL)
+	if err != nil {
+		errMsg := fmt.Sprintf("failed while performing the check due to %v", err)
```

Comment:
is this helpful. does err have e.g. the path? Would also follow the normal format with colon: `.... the check: %v`